src/triage/component/architect/protected_groups_generators.py

```diff
@@ -0,0 +1,125 @@
+            len(as_of_dates)
+        )
+        for as_of_date in as_of_dates:
+            if not self.replace:
+                logging.info(
+                    "Looking for existing protected_groups for as of date %s",
+                    as_of_date
+                )
+                any_existing_rows = list(self.db_engine.execute(
+                    """select 1 from {protected_groups_table}
+                    where as_of_date = '{as_of_date}'
+                    limit 1
+                    """.format(
+                        protected_groups_table=self.protected_groups_table_name,
+                        as_of_date=as_of_date
+                    )
+                ))
+                if len(any_existing_rows) == 1:
+                    logging.info("Since nonzero existing protected_groups found, skipping")
+                    continue
+
+            logging.info(
+                "Generating protected groups for as of date %s ",
+                as_of_date
+            )
+            self.generate(start_date=as_of_date, cohort_table_name=cohort_table_name)
+        nrows = [
+            row[0]
+            for row in self.db_engine.execute(
+                "select count(*) from {}".format(self.protected_groups_table_name)
+            )
+        ][0]
+        if nrows == 0:
+            logging.warning("Done creating protected_groups, but no rows in protected_groups table!")
+        else:
+            logging.info("Done creating protected_groups table %s: rows: %s", self.protected_groups_table_name, nrows)
+
+
+    def generate(self, start_date, cohort_table_name):
+        full_insert_query = textwrap.dedent(
+            """
+            insert into {protected_groups_table}
+            select distinct on (cohort.entity_id, cohort.as_of_date)
+                cohort.entity_id,
+                '{as_of_date}'::date as as_of_date,
+                {attribute_columns} 
+            from {cohort_table_name} cohort 
+            left join (select * from {from_obj}) from_obj  on 
+                cohort.entity_id = from_obj.{entity_id_column} and
+                cohort.as_of_date > from_obj.{knowledge_date_column}
+            where cohort.as_of_date = '{as_of_date}'::date
+            order by cohort.entity_id, cohort.as_of_date, {knowledge_date_column} desc
+        """
+        ).format(
+            protected_groups_table=self.protected_groups_table_name,
+            as_of_date=start_date,
+            attribute_columns=", ".join([str(col) for col in self.attribute_columns]),
+            cohort_table_name=cohort_table_name,
+            from_obj=self.from_obj,
+            knowledge_date_column=self.knowledge_date_column,
+            entity_id_column=self.entity_id_column
+        )
+        logging.debug("Running protected_groups creation query")
+        logging.debug(full_insert_query)
+        self.db_engine.execute(full_insert_query)
+
+    def clean_up(self, protected_groups_table_name):
```

Comment:
This isn't being called anywhere, you could remove it (also the one in the NoOp)